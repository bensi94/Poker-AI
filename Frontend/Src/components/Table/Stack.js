import React from 'react';
import PropTypes from 'prop-types';

import BlackChip from '../../../assets/black-chip-horizontal.svg';
import BlueChip from '../../../assets/blue-chip-horizontal.svg';
import GreenChip from '../../../assets/green-chip-horizontal.svg';
import PinkChip from '../../../assets/pink-chip-horizontal.svg';
import RedChip from '../../../assets/red-chip-horizontal.svg';

class Stack extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            widthRatio: 0.027,
            leftRatio: 0.028,
            topRatio: 0.003,
            heightRatio: 0.025,
            fontRatio: 0.011,
            chipRankList: [
                'PinkChip',
                'BlackChip',
                'RedChip',
                'BlueChip',
                'GreenChip'
            ],
            chipStack: undefined,
            stackWidth: 0
        };
    }

    componentDidMount() {
        this.setState({
            chipStack: this.getChipStack(),
            height: this.state.heightRatio * this.props.tableSize
        });
    }

    getChipComponent(component, style, keyString) {
        switch (component) {
            case 'GreenChip':
                return (<GreenChip
                    style={style}
                    className="horizontal-chip"
                    width={this.props.tableSize * this.state.widthRatio}
                    key={keyString} />
                );
            case 'RedChip':
                return (<RedChip
                    style={style}
                    className="horizontal-chip"
                    width={this.props.tableSize * this.state.widthRatio}
                    key={keyString} />
                );
            case 'BlueChip':
                return (<BlueChip
                    style={style}
                    className="horizontal-chip"
                    width={this.props.tableSize * this.state.widthRatio}
                    key={keyString} />
                );
            case 'PinkChip':
                return (<PinkChip
                    style={style}
                    className="horizontal-chip"
                    width={this.props.tableSize * this.state.widthRatio}
                    key={keyString} />
                );
            case 'BlackChip':
                return (<BlackChip
                    style={style}
                    className="horizontal-chip"
                    width={this.props.tableSize * this.state.widthRatio}
                    key={keyString}/>
                );
            default:
                break;
        }
    }

    getChipStack() {
        let chips = [];
        // We reverse the string to start with the smallest number(chip)
        const amount = this.props.stackAmount.toString().split('').reverse().join('');
        // If chips go above 100k we do not generate them at this point
        for (let i = 0; i < amount.length && i < this.state.chipRankList.length; i++) {
            const currentChipCount = amount[i];
            for (let j = 0; j < currentChipCount; j++) {
                const style = {
                    left: `${this.state.leftRatio * this.props.tableSize * i}px`,
                    top: `-${this.state.topRatio * this.props.tableSize * j}px`,
                    zIndex: j
                };
                chips.push(this.getChipComponent(this.state.chipRankList[i],
                    style, `${i.toString()}_${j.toString()}`));
            }
            this.setState({
                stackWidth: this.state.leftRatio * this.props.tableSize * (i + 1)
            });
        }

        return chips;
    }

    render() {
        let style = {
            left: `${this.props.left}%`,
            top: `${this.props.top}%`,
            transform: `translateX(-${this.state.stackWidth / 2}px)`
        };

        let labelPosClass = this.props.top > 50 ? 'lbl-pos-top' : 'lbl-pos-bottom';

        return (
            <div className="stack-wrapper" style={style}>
                {this.state.chipStack}
                <div className={`bet-amount-lbl ${labelPosClass}`}
                    style={{ fontSize: `${this.props.tableSize * this.state.fontRatio}pt` }}
                >
                    ${ this.props.stackAmount }
                </div>
            </div>
        );
    }
}

Stack.propTypes = {
    stackAmount: PropTypes.number.isRequired,
    tableSize: PropTypes.number.isRequired,
    left: PropTypes.number.isRequired,
    top: PropTypes.number.isRequired
};

export default Stack;
