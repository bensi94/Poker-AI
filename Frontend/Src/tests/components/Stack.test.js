import React from 'react';
import { shallow } from 'enzyme';

import Stack from '../../components/Table/Stack';


describe('Stack Test suite', () => {
    let wrapper;

    it('Should render component correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('Should Generate correct stack for amount: 2543', () => {
        wrapper = shallow(
            <Stack stackAmount={2543}
                tableSize={500}
                left={10}
                top={10}/>);
        // 3 + 4 + 5 + 2 = 14
        expect(wrapper.instance().getChipStack().length).toBe(14);
    });

    it('Should Generate correct stack for amount: 24839', () => {
        wrapper = shallow(
            <Stack stackAmount={24839}
                tableSize={500}
                left={10}
                top={10}/>);
        // 9 + 3 + 8 + 4 + 2 = 26
        expect(wrapper.instance().getChipStack().length).toBe(26);
    });

    it('Should be calculating the style(positions correctly in getChipStack()', () => {
        wrapper = shallow(
            <Stack stackAmount={12}
                tableSize={1000}
                left={10}
                top={10}/>);

        wrapper.instance().setState({
            widthRatio: 1,
            leftRatio: 1,
            topRatio: 1,
            heightRatio: 1
        });

        const responseList = wrapper.instance().getChipStack();
        const respOne = responseList[0];
        const respTwo = responseList[1];
        const respThree = responseList[2];

        const styleOne = {
            left: '0px',
            top: '-0px',
            zIndex: 0
        };
        const styleTwo = {
            left: '0px',
            top: '-1000px',
            zIndex: 1
        };
        const styleThree = {
            left: '1000px',
            top: '-0px',
            zIndex: 0
        };

        // First component
        expect(respOne.props.width).toBe(1000);
        expect(respOne.props.style).toStrictEqual(styleOne);

        // Second component
        expect(respTwo.props.width).toBe(1000);
        expect(respTwo.props.style).toStrictEqual(styleTwo);

        // Third component
        expect(respThree.props.width).toBe(1000);
        expect(respThree.props.style).toStrictEqual(styleThree);
    });
});
