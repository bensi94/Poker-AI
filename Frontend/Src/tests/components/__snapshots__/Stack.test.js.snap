// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack Test suite Should render component correctly 1`] = `
<div
  className="stack-wrapper"
  style={
    Object {
      "left": "10%",
      "top": "10%",
      "transform": "translateX(-35px)",
    }
  }
>
  <stubbed-asset
    className="horizontal-chip"
    key="0_0"
    style={
      Object {
        "left": "0px",
        "top": "-0px",
        "zIndex": 0,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="0_1"
    style={
      Object {
        "left": "0px",
        "top": "-1.5px",
        "zIndex": 1,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="0_2"
    style={
      Object {
        "left": "0px",
        "top": "-3px",
        "zIndex": 2,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="0_3"
    style={
      Object {
        "left": "0px",
        "top": "-4.5px",
        "zIndex": 3,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="0_4"
    style={
      Object {
        "left": "0px",
        "top": "-6px",
        "zIndex": 4,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="0_5"
    style={
      Object {
        "left": "0px",
        "top": "-7.5px",
        "zIndex": 5,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="0_6"
    style={
      Object {
        "left": "0px",
        "top": "-9px",
        "zIndex": 6,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="0_7"
    style={
      Object {
        "left": "0px",
        "top": "-10.5px",
        "zIndex": 7,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="0_8"
    style={
      Object {
        "left": "0px",
        "top": "-12px",
        "zIndex": 8,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="1_0"
    style={
      Object {
        "left": "14px",
        "top": "-0px",
        "zIndex": 0,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="1_1"
    style={
      Object {
        "left": "14px",
        "top": "-1.5px",
        "zIndex": 1,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="1_2"
    style={
      Object {
        "left": "14px",
        "top": "-3px",
        "zIndex": 2,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="2_0"
    style={
      Object {
        "left": "28px",
        "top": "-0px",
        "zIndex": 0,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="2_1"
    style={
      Object {
        "left": "28px",
        "top": "-1.5px",
        "zIndex": 1,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="2_2"
    style={
      Object {
        "left": "28px",
        "top": "-3px",
        "zIndex": 2,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="2_3"
    style={
      Object {
        "left": "28px",
        "top": "-4.5px",
        "zIndex": 3,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="2_4"
    style={
      Object {
        "left": "28px",
        "top": "-6px",
        "zIndex": 4,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="2_5"
    style={
      Object {
        "left": "28px",
        "top": "-7.5px",
        "zIndex": 5,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="2_6"
    style={
      Object {
        "left": "28px",
        "top": "-9px",
        "zIndex": 6,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="2_7"
    style={
      Object {
        "left": "28px",
        "top": "-10.5px",
        "zIndex": 7,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="3_0"
    style={
      Object {
        "left": "42px",
        "top": "-0px",
        "zIndex": 0,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="3_1"
    style={
      Object {
        "left": "42px",
        "top": "-1.5px",
        "zIndex": 1,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="3_2"
    style={
      Object {
        "left": "42px",
        "top": "-3px",
        "zIndex": 2,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="3_3"
    style={
      Object {
        "left": "42px",
        "top": "-4.5px",
        "zIndex": 3,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="4_0"
    style={
      Object {
        "left": "56px",
        "top": "-0px",
        "zIndex": 0,
      }
    }
    width={13.5}
  />
  <stubbed-asset
    className="horizontal-chip"
    key="4_1"
    style={
      Object {
        "left": "56px",
        "top": "-1.5px",
        "zIndex": 1,
      }
    }
    width={13.5}
  />
  <div
    className="bet-amount-lbl lbl-pos-bottom"
    style={
      Object {
        "fontSize": "5.5pt",
      }
    }
  >
    $
    24839
  </div>
</div>
`;
